tests/TestCase/Database/ExpressionTypeCastingTest.php

```diff
@@ -0,0 +1,171 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The Open Group Test Suite License
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.3.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Test\TestCase\Database;
+
+use Cake\TestSuite\TestCase;
+use Cake\Database\Type\StringType;
+use Cake\Database\Type\ExpressionTypeInterface;
+use Cake\Database\Expression\FunctionExpression;
+use Cake\Database\ValueBinder;
+use Cake\Database\Type;
+use Cake\Database\Expression\Comparison;
+use Cake\Database\Expression\BetweenExpression;
+use Cake\Database\Expression\CaseExpression;
+
+class TestType extends StringType implements ExpressionTypeInterface
+{
+
+    public function toExpression($value)
+    {
+        return new FunctionExpression('CONCAT', [$value, ' - foo']);
+    }
+}
+
+/**
+ * Tests for Expression objects casting values to other expressions
+ * using the type classes
+ *
+ */
+class ExpressionTypeCastingTest extends TestCase
```

Comment:
Another case that would be good to cover is a use case of a Type class that maps a column to a Value object and then back into an expression. I know that was one of the big asks in #8041